staging/src/k8s.io/apiserver/pkg/endpoints/discovery/v1/handler.go

```diff
@@ -0,0 +1,265 @@
+/*
+Copyright 2016 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package v1
+
+import (
+	"net/http"
+	"sync"
+
+	"github.com/emicklei/go-restful/v3"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apiserver/pkg/endpoints/handlers/negotiation"
+	"k8s.io/klog/v2"
+)
+
+const DiscoveryEndpointRoot = "/discovery"
+
+// This handler serves the /discovery/v1 endpoint for a given list of
+// api resources indexed by their group version.
+type ResourceManager interface {
+	// Adds knowledge of the given groupversion to the discovery document
+	// If it was already being tracked, updates the stored DiscoveryGroupVersion
+	// Thread-safe
+	AddGroupVersion(groupName string, value metav1.DiscoveryGroupVersion)
+
+	// Removes all group versions for a given group
+	// Thread-safe
+	RemoveGroup(groupName string)
+
+	// Removes a specific groupversion. If all versions of a group have been
+	// removed, then the entire group is unlisted.
+	// Thread-safe
+	RemoveGroupVersion(gv metav1.GroupVersion)
+
+	// Reset's from the manager known list of group-versions and replaces it
+	// with the given groups
+	// Thread-Safe
+	SetGroups([]metav1.DiscoveryAPIGroup)
+
+	// Returns a restful webservice which responds to discovery requests
+	// Thread-safe
+	WebService() *restful.WebService
+
+	http.Handler
+}
+
+type resourceDiscoveryManager struct {
+	// Protects writes to all fields in struct
+	lock sync.RWMutex
+
+	// Writes protected by the lock. List if all apigroups & resources indexed
+	// by the resource manager
+	apiGroups     map[string]metav1.DiscoveryAPIGroup
+	apiGroupNames []string // apiGroupNames preserves insertion order
+
+	serializer runtime.NegotiatedSerializer
+
+	// Most up to date version of all discovery api groups
+	cachedResponse *metav1.DiscoveryAPIGroupList
+	// Hash of the cachedResponse used for cache-busting
+	cachedResponseETag string
+}
+
+func NewResourceManager(serializer runtime.NegotiatedSerializer) ResourceManager {
+	result := &resourceDiscoveryManager{serializer: serializer}
+	return result
+}
+
+func (rdm *resourceDiscoveryManager) SetGroups(groups []metav1.DiscoveryAPIGroup) {
+	rdm.lock.Lock()
+	defer rdm.lock.Unlock()
+
+	rdm.apiGroups = nil
+	rdm.cachedResponse = nil
```

Comment:
why isn't the cachedResponseETag cleared?

Reply:
it is now being cleared